usagerep/usagerep.go

```diff
@@ -28,39 +27,104 @@ import (
 
 	"github.com/projectcalico/felix/buildinfo"
 	"github.com/projectcalico/felix/calc"
+	"github.com/projectcalico/felix/jitter"
 )
 
 const (
-	baseURL = "https://usage.projectcalico.org/UsageCheck/calicoVersionCheck?"
+	DefaultBaseURL = "https://usage.projectcalico.org/UsageCheck/calicoVersionCheck?"
 )
 
-func PeriodicallyReportUsage(interval time.Duration, clusterGUID, clusterType, calicoVersion string, statsUpdateC <-chan calc.StatsUpdate) {
-	log.Info("Usage reporting thread started, waiting for size estimate")
-	stats := <-statsUpdateC
-	log.WithField("stats", stats).Info("Initial stats read")
+func New(
+	interval time.Duration,
+	statsUpdateC <-chan calc.StatsUpdate,
+	configUpdateC <-chan map[string]string,
+) *UsageReporter {
+	return &UsageReporter{
+		interval:      interval,
+		statsUpdateC:  statsUpdateC,
+		configUpdateC: configUpdateC,
+		InitialDelay:  5 * time.Minute,
+		BaseURL:       DefaultBaseURL,
+	}
+}
+
+type UsageReporter struct {
+	interval      time.Duration
+	statsUpdateC  <-chan calc.StatsUpdate
+	configUpdateC <-chan map[string]string
+
+	InitialDelay time.Duration
+	BaseURL      string
+}
 
-	// To avoid thundering herd, inject some startup jitter.
-	initialDelay := calculateInitialDelay(stats.NumHosts)
-	log.WithField("delay", initialDelay).Info("Waiting before first check-in")
-	time.Sleep(initialDelay)
+func (u *UsageReporter) PeriodicallyReportUsage(ctx context.Context) {
+	var stats calc.StatsUpdate
+	var config map[string]string
+	var receivedFirstStats bool
+	var tickerC <-chan time.Time
+	initialDelayStarted := false
+	initialDelayDone := make(chan struct{})
 
-	log.Info("Initial delay complete, making first check-in")
-	ReportUsage(clusterGUID, clusterType, calicoVersion, stats)
+	maybeStartInitialDelay := func() {
+		if !receivedFirstStats {
+			return
+		}
+		if config == nil {
+			return
+		}
+		if initialDelayStarted {
+			return
+		}
+
+		// To avoid thundering herd, inject some startup jitter.
+		initialDelay := u.calculateInitialDelay(stats.NumHosts)
+		func() {
```

Comment:
Should this be in a goroutine?  Otherwise looks like this will block the main loop below.

Reply:
Yes!